hash_map.h

```diff
@@ -0,0 +1,197 @@
+#pragma once
+#include <iostream>
+#include <vector>
+#include <list>
+#include <stdexcept>
+#include <utility>
+
+template<
+  class KeyType,
+  class ValueType,
+  class Hash = std::hash<KeyType>>
+class HashMap {
+ private:
+  using size_type = size_t;
+  using list_type = std::list<std::pair<const KeyType, ValueType>>;
+  using table_type = std::vector<std::list<typename list_type::iterator>>;
+
+  Hash hasher_;
+
+  list_type list_of_pairs_;
+  table_type table_;
+
+  const size_type kMinSize = 16;
+  const double kAlpha = 0.5;
+  size_type elements_stored_;
+  size_type slots_available_;
+
+  void rehash() {
+    slots_available_ *= 2;
+    table_type table_copy(slots_available_, { list_of_pairs_.end() });

```

Comment:
With this table, you will copy all iterators twice. Once on rehash and once, on `table_ = table_copy`, please copy only once. 

`table_.clear();` will help you.